File: server/js/src/v2/adapters/langchain.ts
/**
 * Adapter for LangChain.js chat models (ChatOpenAI, ChatAnthropic,
 * ChatGoogleGenerativeAI, ...) — anything exposing `.stream()` and yielding
 * AIMessageChunk. Provider-agnostic: swap the underlying chat model
 * without touching this adapter.
 *
 *   const llm = new ChatOpenAI({ model: "gpt-4o" }).bindTools([
 *     mzLangchain.makeAdFactsTool(session),
 *   ]);
 *   const chunks = await llm.stream(messages);
 *   for await (const event of session.stream(mzLangchain.toTextStream(chunks), decision)) { ... }
 */
import { MonetzlySession } from "../session.js";
import { AD_FACTS_TOOL_DESCRIPTION, AD_FACTS_TOOL_NAME, adFactsPayload } from "./common.js";

interface AIMessageChunkLike {
  content: string | Array<{ type?: string; text?: string }>;
}

/**
 * chunks: the async iterable from `llm.stream(messages)`. Yields text
 * content off each AIMessageChunk (string content, or the "text" parts of
 * multimodal content arrays); chunks carrying only tool_call_chunks are
 * skipped.
 */
export async function* toTextStream(
  chunks: AsyncIterable<AIMessageChunkLike>,
): AsyncGenerator<string> {
  for await (const chunk of chunks) {
    const { content } = chunk;
    if (typeof content === "string") {
      if (content) yield content;
    } else if (Array.isArray(content)) {
      for (const part of content) {
        if (part.type === "text" && part.text) yield part.text;
      }
    }
  }
}

/**
 * Returns a LangChain DynamicStructuredTool, bindable via
 * `llm.bindTools([makeAdFactsTool(session)])`. Imports @langchain/core and
 * zod lazily so this module has no hard dependency on them otherwise.
 */
export async function makeAdFactsTool(session: MonetzlySession) {
  const { tool } = await import("@langchain/core/tools");
  const { z } = await import("zod");

  return tool(
    async ({ ad_id }: { ad_id: string }) => JSON.stringify(await adFactsPayload(session, ad_id)),
    {
      name: AD_FACTS_TOOL_NAME,
      description: AD_FACTS_TOOL_DESCRIPTION,
      schema: z.object({
        ad_id: z.string().describe("The ad id from the history annotation"),
      }),
    },
  );
}

File: server/python/monetzly/v2/adapters/langchain.py
"""
Adapter for LangChain chat models (ChatOpenAI, ChatAnthropic,
ChatGoogleGenerativeAI, ...) — anything exposing `.astream()` and yielding
AIMessageChunk. Provider-agnostic: swap the underlying chat model without
touching this adapter.

    from monetzly.v2.adapters import langchain as mz_lc

    llm = ChatOpenAI(model="gpt-4o", streaming=True).bind_tools(
        [mz_lc.make_ad_facts_tool(session)])
    chunks = llm.astream(messages)
    async for event in session.stream(mz_lc.to_text_stream(chunks), decision):
        ...
"""
from typing import Any, AsyncIterator

from ._common import AD_FACTS_TOOL_DESCRIPTION, AD_FACTS_TOOL_NAME, ad_facts_payload


async def to_text_stream(chunks: Any) -> AsyncIterator[str]:
    """chunks: the async iterator from `llm.astream(messages)`. Yields
    `.content` off each AIMessageChunk; chunks carrying only
    tool_call_chunks are skipped (LangChain surfaces the accumulated tool
    call once bind_tools is set — read it off the final chunk, or use
    `llm.bind_tools(...).ainvoke(...)` instead of streaming if you need the
    call before any text)."""
    async for chunk in chunks:
        content = getattr(chunk, "content", None)
        if content:
            yield content


def make_ad_facts_tool(session):
    """Returns a LangChain StructuredTool, bindable via
    `llm.bind_tools([make_ad_facts_tool(session)])`. Imports langchain-core
    lazily so this module has no hard dependency on it otherwise."""
    from langchain_core.tools import StructuredTool
    from pydantic import BaseModel, Field

    class AdFactsInput(BaseModel):
        ad_id: str = Field(description="The ad id from the history annotation")

    async def _run(ad_id: str) -> dict:
        return await ad_facts_payload(session, ad_id)

    return StructuredTool.from_function(
        coroutine=_run,
        name=AD_FACTS_TOOL_NAME,
        description=AD_FACTS_TOOL_DESCRIPTION,
        args_schema=AdFactsInput,
    )
